executor/executor_runner.h

```diff
@@ -0,0 +1,780 @@
+		// with fork server the top process can exit with kFailStatus if it wants special handling.
+		if (status != kFailStatus)
+			status = 0;
+		if (FailCurrentRequest(status == kFailStatus)) {
+			// Read out all pening output until EOF.
+			if (IsSet(msg_->flags, rpc::RequestFlag::ReturnOutput)) {
+				while (ReadOutput())
+					;
+			}
+			HandleCompletion(status);
+		} else if (attempts_ > 3)
+			sleep_ms(100 * attempts_);
+		Start();
+	}
+
+	bool FailCurrentRequest(bool failed)
+	{
+		if (state_ == State::Handshaking)
+			return failed && IsSet(msg_->flags, rpc::RequestFlag::ReturnError);
+		if (state_ == State::Executing)
+			return !failed || IsSet(msg_->flags, rpc::RequestFlag::ReturnError);
+		return false;
+	}
+
+	void Start()
+	{
+		state_ = State::Started;
+		freshness_ = 0;
+		int req_pipe[2];
+		if (pipe(req_pipe))
+			fail("pipe failed");
+		int resp_pipe[2];
+		if (pipe(resp_pipe))
+			fail("pipe failed");
+		int stdout_pipe[2];
+		if (pipe(stdout_pipe))
+			fail("pipe failed");
+
+		std::vector<std::pair<int, int>> fds = {
+		    {req_pipe[0], STDIN_FILENO},
+		    {resp_pipe[1], STDOUT_FILENO},
+		    {stdout_pipe[1], STDERR_FILENO},
+		    {req_shmem_.FD(), kInFd},
+		    {resp_shmem_.FD(), kOutFd},
+		    {max_signal_fd_, kMaxSignalFd},
+		    {cover_fitler_fd_, kCoverFilterFd},
+		};
+		const char* argv[] = {bin_, "exec", nullptr};
+		process_.emplace(argv, fds);
+
+		Select::Prepare(resp_pipe[0]);
+		Select::Prepare(stdout_pipe[0]);
+
+		close(req_pipe[0]);
+		close(resp_pipe[1]);
+		close(stdout_pipe[1]);
+
+		close(req_pipe_);
+		close(resp_pipe_);
+		close(stdout_pipe_);
+
+		req_pipe_ = req_pipe[1];
+		resp_pipe_ = resp_pipe[0];
+		stdout_pipe_ = stdout_pipe[0];
+
+		if (msg_)
+			Handshake();
+	}
+
+	void Handshake()
+	{
+		if (state_ != State::Started || !msg_)
+			fail("wrong handshake state");
+		debug("proc %d: handshaking to execute request %llu\n", id_, static_cast<uint64>(msg_->id));
+		state_ = State::Handshaking;
+		exec_start_ = current_time_ms();
+		exec_env_ = msg_->exec_opts->env_flags() & ~rpc::ExecEnv::ResetState;
+		sandbox_arg_ = msg_->exec_opts->sandbox_arg();
+		handshake_req req = {
+		    .magic = kInMagic,
+		    .flags = exec_env_,
+		    .pid = static_cast<uint64>(id_),
+		    .sandbox_arg = static_cast<uint64>(sandbox_arg_),
+		};
+		if (write(req_pipe_, &req, sizeof(req)) != sizeof(req)) {
+			debug("request pipe write failed (errno=%d)\n", errno);
+			Restart();
+		}
+	}
+
+	void Execute()
+	{
+		if (state_ != State::Idle || !msg_)
+			fail("wrong state for execute");
+
+		debug("proc %d: start executing request %llu\n", id_, static_cast<uint64>(msg_->id));
+
+		rpc::ExecutingMessageRawT exec;
+		exec.id = msg_->id;
+		exec.proc_id = id_;
+		exec.try_ = attempts_;
+
+		if (wait_start_) {
+			exec.wait_duration = (wait_end_ - wait_start_) * 1000 * 1000;
+			wait_end_ = wait_start_ = 0;
+		}
+
+		rpc::ExecutorMessageRawT raw;
+		raw.msg.Set(std::move(exec));
+		conn_.Send(raw);
```

Comment:
In the end, who frees the memory allocated for the raw object? Do flatbuffers take care of destruction/freeing themselves?